File: lclang/src/main/java/lclang/types/NamedType.java
package lclang.types;

import lclang.libs.lang.classes.LCClass;

public class NamedType extends Type {
    public final LCClass clazz;
    public NamedType(LCClass clazz) {
        super(clazz.name);
        this.clazz = clazz;
    }

    @Override
    public boolean isAcceptWithoutNullable(Type another) {
        return super.isAcceptWithoutNullable(another) || (
                another instanceof NamedType &&
                ((NamedType) another).clazz.canCast(clazz));
    }
}

File: lclang/src/main/kotlin/lclang/types/NamedType.kt
package lclang.types

import lclang.LCClass

open class NamedType(
    val clazz: LCClass,
): Type(clazz.name) {
    override fun isAcceptWithoutNullable(another: Type): Boolean {
        return super.isAcceptWithoutNullable(another) ||
                        another is Types.MagicType &&
                        another.text == clazz.name ||
                another is NamedType &&
                another.clazz.canCast(clazz)
    }
}